File: src/leetCode/Ex57.java
package leetCode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Ex57 {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        int left = newInterval[0];
        int right = newInterval[1];
        boolean placed = false;
        List<int[]> ansList = new ArrayList<int[]>();
        for (int[] interval : intervals) {
            if(left>interval[1]){
                ansList.add(interval);
            }else if(right<interval[0]){
                if(!placed){
                    ansList.add(new int[]{left, right});
                    placed = true;
                }
                ansList.add(interval);
            }else{
                left = Math.min(left, interval[0]);
                right = Math.max(right, interval[1]);
            }
        }
        if(!placed){
            ansList.add(new int[]{left, right});
        }
        int[][] res = new int[ansList.size()][];
        for (int i = 0; i < res.length; i++) {
            res[i] = ansList.get(i);
        }
        return res;
    }


    public static void main(String[] args) {
        var o = new Ex57();
        var res = o.insert(new int[][]{{1,3},{6,9}}, new int[]{2,5});
        for (int[] is : res) {
            System.out.println(Arrays.toString(is));
        }
    }
}

File: src/leetCode/ex57.kt
package leetCode

fun insert(intervals: List<Interval>, newInterval: Interval) = merge2(mutableListOf<Interval>().apply {
    addAll(intervals);add(newInterval)
})
private fun merge2(intervals: List<Interval>): List<Interval> {
    if (intervals.isEmpty()) return listOf()
    fun mergeIntervals(l: Interval, r: Interval) {
        if (l.end < r.end)
            l.end = r.end
    }
    //initialization a stack through kotlin hasn't contain
    val res = mutableListOf<Interval>()
    val int = intervals.sortedBy { it.start }
    res.add(int[0])
    (1 until int.size).forEach { index ->
        when {
            res[res.size - 1].end >= int[index].start -> mergeIntervals(res[res.size - 1], int[index])
            else -> res.add(int[index])
        }
    }
    return res
}

fun main(args: Array<String>) {
    insert(listOf(Interval(1,5)),Interval(2,3)).forEach { println("${it.start} : ${it.end}") }
}